### PR TITLE
wasi-threads: an initial implementation

### PR DESCRIPTION
**Pull request by abrown**
This PR builds on several other under-review PRs to demonstrate a working wasi-threads implementation in Wasmtime. The first three commits are cherry-picked from the following PRs:
 - #5326 
 - #5471
 - #5475 

The last two commits build on these to expose the wasi-threads API to WebAssembly modules. Upon building and running Wasmtime with the right flags, one can then run multi-threaded WebAssembly programs:

```console
$ cargo run --features wasi-threads -- \
    --wasm-features threads --wasi-modules experimental-wasi-threads \
    <a threads-enabled module>
```

Due to Cargo feature issues in 05020d4, this PR is not yet expected to pass all CI tasks but, perhaps more interestingly, one can use these changes to run some simple pthread tests over in the wasi-libc repository: https://github.com/WebAssembly/wasi-libc/pull/369.

### CODE REVIEW

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 1]
```diff
@@ -1115,22 +1112,18 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
             // `&mut [u8]`'s exclusive ownership which is not guaranteed due to
             // potential access from other threads).
             let mut tmp = vec![0; buf.len().min(MAX_SHARED_BUFFER_SIZE)];
-            self.random.try_fill_bytes(&mut tmp)?;
+            self.random.lock().unwrap().try_fill_bytes(&mut tmp)?;
```
Comment (alexcrichton) [inline crates/wasi-common/src/snapshots/preview_1.rs:1115]: Could you file an issue and list a TODO here for "we shouldn't serialize calls to `random_get`"?

File: crates/wasi-threads/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# wasmtime-wasi-threads
```
Comment (alexcrichton) [inline crates/wasi-threads/README.md:1]: It's ok to leave this file out

File: crates/wasi-threads/LICENSE [hunk 3]
```diff
@@ -0,0 +1,220 @@
+
```
Comment (alexcrichton) [inline crates/wasi-threads/LICENSE:1]: If possible I think we're omitting LICENSE files except for the one at the top of the repo

File: tests/all/cli_tests/threads.wat [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+
+    ;; Wasmtime has no `wait/notify` yet, so we just spin to allow the threads
+    ;; to do their work.
+    (local.set $i (i32.const 2000000))
+    (loop $again
+      (local.set $i (i32.sub (local.get $i) (i32.const 1)))
+      (br_if $again (i32.gt_s (local.get $i) (i32.const 0)))
+    )
```
Comment (alexcrichton) [inline tests/all/cli_tests/threads.wat:27]: Technically this can be done today, but I think this might be best done as some sort of "join" API from the host which isn't currently specified for `wasi-threads`, so while wait/notify could be added here there's no way to actually wait until all the child threads have exited.

File: crates/wasi-threads/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
+
+    pub fn spawn(&self, host: T, thread_start_arg: i32) -> Result<i32> {
+        let module = self.module.clone();
+        let linker = self.linker.clone();
```
Comment (alexcrichton) [inline crates/wasi-threads/src/lib.rs:27]: If possible I think this would be best avoided because a `Linker<T>` is a costly thing to clone.

One option is to wrap it up in an `Arc`, and another would be to thread the `get_cx` callback through to `spawn` so the `host` argument could be projected into a `&WasiThreadsCtx<T>`

File: crates/wasi-threads/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            let mut store = Store::new(&module.engine(), host);
+            let instance = linker
+                .instantiate(&mut store, &module)
+                .expect(&fail("failed to instantiate".into()));
+            let thread_entry_point = instance
+                .get_typed_func::<(i32, i32), ()>(&mut store, WASI_ENTRY_POINT)
+                .expect(&fail(format!(
+                    "failed to find wasi-threads entry point function: {}",
+                    WASI_ENTRY_POINT
+                )));
```
Comment (alexcrichton) [inline crates/wasi-threads/src/lib.rs:52]: I feel like this is a little tricky. Panicking on errors here doesn't feel like the right thing to do since these panics can theoreteically get triggered. In that sense I think that the errors need to be handled here. As you've probably noticed, however, there's nowhere to communicate the error to. One option would be to move the initialization here to the thread calling spawn, but that feels sort of wrong since it sort of belongs in the child thread.

Overall I think this might be best left as an issue on the `wasi-threads` repository. Basically how would a thread spawn operation get notified that the thread was spawned but something else failed? Or otherwise how are traps handled and/or other fatal errors in the threads proposal? (that is the `panic!` below also needs better handling I think)

File: crates/wasi-common/src/snapshots/preview_1.rs (hunk 1 above)
Comment (abrown) [reply crates/wasi-common/src/snapshots/preview_1.rs]: It felt strange to create an issue for yet-to-be-merged code; I will do this as a follow-up.

File: crates/wasi-threads/src/lib.rs (hunk 6 above)
Comment (abrown) [reply crates/wasi-threads/src/lib.rs:52]: Yeah, I've looked into this more since you reviewed this. There has been some discussion [in a wasi-threads issue](https://github.com/WebAssembly/wasi-threads/pull/17/files) which boils down to "any `proc_exit` or trap in any thread exits the program." I was looking at [what `run.rs` does](https://github.com/bytecodealliance/wasmtime/blob/main/src/commands/run.rs#L215-L246) for traps and I think we would want to do something similar here--essentially call `std::process::exit`. The problem is that calling `std::process::exit` is likely not what one wants to do in an embedding scenario. I would suggest we do implement wasi-threads to initially `std::process::exit` and then open an issue to think about this for the embedding scenario: I suspect users here would want a trap or `proc_exit` to kill all the threads but not kill the original program calling Wasmtime but, with no great way to cancel threads in Rust, I think this might be kind of difficult.

Comment (alexcrichton) [reply crates/wasi-threads/src/lib.rs:52]: I'm ok with this all being a TODO and not necessarily covered here. At the very least this PR is a good step forward to being able to spawn threads.

I do think these issues need to be resolved before stabilization, but perhaps it's best to track these sorts of things in the wasi-threads repository?

Comment (abrown) [reply crates/wasi-threads/src/lib.rs:52]: Ok, to resolve this I rebased again on `main` and applied [2ba9f11](https://github.com/bytecodealliance/wasmtime/pull/5484/commits/2ba9f11b56adaef32f5bc2596cafab8328e52ae4).

File: tests/all/cli_tests/threads.wat (hunk 4 above)
Comment (abrown) [reply tests/all/cli_tests/threads.wat:27]: Looking at this again, I think I can rewrite this with a "join" loop after the spawn calls: each thread will atomically increment a memory location and notify the main thread; the main thread waits in a loop and only exits once the read memory equals the number of threads we are waiting for. This fixup could come after this PR merges?

File: crates/wasi-common/cap-std-sync/src/file.rs [hunk 7]
```diff
@@ -14,11 +15,48 @@ use wasi_common::{
...
+
+#[cfg(windows)]
+impl AsRawHandleOrSocket for BorrowedFile<'_> {
+    #[inline]
+    fn as_raw_handle_or_socket(&self) -> RawHandleOrSocket {
+        self.0.as_raw_handle_or_socket()
+    }
+}
+
+pub struct File(RwLock<cap_std::fs::File>);
```
Comment (alexcrichton) [inline crates/wasi-common/cap-std-sync/src/file.rs:51]: Oh I had forgotten to mention this earlier, but now that other bits and pieces are cleaned up I think most of these changes should no longer be necessary? I think the `BorrowedFile` type should be able to be removed now along with all other changes necessary for this.

File: crates/wasi-common/cap-std-sync/src/sched/unix.rs [hunk 8]
```diff
@@ -8,28 +8,33 @@ pub async fn poll_oneoff<'a>(poll: &mut Poll<'a>) -> Result<(), Error> {
...
+                poll_as_fds.push((fd, PollFlags::OUT));
             }
             Subscription::MonotonicClock { .. } => unreachable!(),
         }
     }
 
+    let mut pollfds = poll_as_fds
+        .iter()
+        .map(|(fd, events)| PollFd::from_borrowed_fd(fd.as_fd(), events.clone()))
+        .collect::<Vec<_>>();
```
Comment (alexcrichton) [inline crates/wasi-common/cap-std-sync/src/sched/unix.rs:36]: Similar to above I think many of these changes can go away now too

File: crates/wasi-common/src/file.rs [hunk 9]
```diff
@@ -317,3 +324,39 @@ pub enum Advice {
...
+    }
+}
+
+#[cfg(windows)]
+impl<T: AsRawHandleOrSocket> AsRawHandleOrSocket for BorrowedAsRawHandleOrSocket<'_, T> {
+    #[inline]
+    fn as_raw_handle_or_socket(&self) -> RawHandleOrSocket {
+        self.0.as_raw_handle_or_socket()
+    }
+}
```
Comment (alexcrichton) [inline crates/wasi-common/src/file.rs:362]: Like with files I think this can all go away now

File: crates/wasi-common/src/ctx.rs [hunk 10]
```diff
@@ -2,16 +2,26 @@ use crate::clocks::WasiClocks;
...
+/// improvements that can be made (TODO:
+/// https://github.com/bytecodealliance/wasmtime/issues/5643).
+#[derive(Clone)]
+pub struct WasiCtx(Arc<WasiCtxInner>);
+
+pub struct WasiCtxInner {
     pub args: StringArray,
     pub env: StringArray,
-    pub random: Box<dyn RngCore + Send + Sync>,
+    pub random: Mutex<Box<dyn RngCore + Send + Sync>>,
```
Comment (alexcrichton) [inline crates/wasi-common/src/ctx.rs:24]: Could you add a `TODO` here that the `Mutex` here shouldn't be necessary? Something along the lines that this forces threads to serialize on their acquisition of randomness but there's need for such serialization.

File: crates/wasi-threads/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Linker<T>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
```
Comment (alexcrichton) [inline crates/wasi-threads/src/lib.rs:23]: There are two unwraps below during the `spawn` operation's closure: 

* Using `Linker::instantiate`
* The call to `get_typed_func`

Both of these, however, can be type-checked ahead of time at this location I believe by inspecting `Module` and `Linker<T>`. Could you either add a TODO to perform such type-checking here or implement it in this PR?

File: crates/wasi-threads/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                WASI_ENTRY_POINT,
+                thread_start_arg
+            );
+            match thread_entry_point.call(&mut store, (wasi_thread_id, thread_start_arg)) {
+                Ok(_) => {}
+                Err(e) => {
+                    log::trace!("exiting thread id = {} due to error", wasi_thread_id);
+                    let e = maybe_exit_on_error(e);
+                    eprintln!("Error: {:?}", e);
+                    std::process::exit(1);
```
Comment (alexcrichton) [inline crates/wasi-threads/src/lib.rs:70]: One possibility that isn't caught by this is the `.call(..)` function panicking, such as trying to use wasi-nn in a threaded context. If a panic happens that leaves all other threads alive and running and just this thread exits. For that it may be worthwhile, for now at least, to catch panics here as well and abort the whole process if one is caught.

File: crates/wasi-common/src/file.rs [hunk 13]
```diff
@@ -317,3 +324,39 @@ pub enum Advice {
     DontNeed,
     NoReuse,
 }
+
+#[cfg(unix)]
+pub struct BorrowedAsFd<'a, T: AsFd>(&'a T);
```
Comment (alexcrichton) [inline crates/wasi-common/src/file.rs:329]: I think this type can be dropped as well? (a cursory look and I'm not seeing anywhere else it's used at least)

File: crates/wasi-threads/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+                        log::trace!("exiting thread id = {} due to error", wasi_thread_id);
+                        let e = maybe_exit_on_error(e);
+                        eprintln!("Error: {:?}", e);
+                        std::process::exit(1);
+                    }
+                }
+            }));
+
+            if let Err(e) = result {
+                eprintln!("Error: {:?}", e);
```
Comment (alexcrichton) [inline crates/wasi-threads/src/lib.rs:83]: Could this perhaps `s/Error/thread panicked/` to clearly indicate a panic happen?

File: src/commands/run.rs [hunk 15]
```diff
@@ -432,19 +413,22 @@ impl RunCommand {
...
+    #[cfg(feature = "wasi-threads")]
+    wasi_threads: Option<std::sync::Arc<WasiThreadsCtx<Host>>>,
 }
 
 /// Populates the given `Linker` with WASI APIs.
 fn populate_with_wasi(
-    store: &mut Store<Host>,
+    host: &mut Host,
     linker: &mut Linker<Host>,
+    _module: Module,
```
Comment (alexcrichton) [inline src/commands/run.rs:431]: Could this be `module` with something like this in the body of the function:

```rust
// silence unused warning as it's only used with wasi-threads
drop(&module);
```

File: src/commands/run.rs [hunk 16]
```diff
@@ -475,7 +459,22 @@ fn populate_with_wasi(
...
+    if wasi_modules.wasi_crypto {
+        #[cfg(not(feature = "wasi-crypto"))]
+        {
+            bail!("Cannot enable wasi-crypto when the binary is not compiled with this feature.");
+        }
+        #[cfg(feature = "wasi-crypto")]
+        {
+            wasmtime_wasi_crypto::add_to_linker(linker, |host| {
+                std::sync::Arc::get_mut(host.wasi_crypto.as_mut().unwrap())
+                    .expect("wasi-crypto is not implemented with multi-threading support")
```
Comment (alexcrichton) [inline src/commands/run.rs:474]: Could the `Arc` type be imported into this module? (it's used in a number of places and should avoid the `std::sync::...` prefix)

Additionally can you leave some comments here for what's going on? Basically something along the lines of how `wasi-crypto` the implementation is not ready for threads yet so the `get_mut` implementation here will succeed only if a thread hasn't been spawned. If a thread is spawned, however, this will panic and indicate to the user that work needs to happen to integrate wasi-crypto and threads.

(and then a comment on `wasi-nn` below referring to this comment as well)

**Review by alexcrichton — APPROVED** [review]

**Comment by alexcrichton** [comment]
Oh one thing as well, it looks like this test may not be running on CI since it requires explicit opt-in to run. That being said including the threads feature in the build by default I think is reasonable since it doesn't pull in any extra build requirements and still requires a flag to enable at runtime.

**Comment by abrown** [comment]
> Oh one thing as well, it looks like this test may not be running on CI since it requires explicit opt-in to run. That being said including the threads feature in the build by default I think is reasonable since it doesn't pull in any extra build requirements and still requires a flag to enable at runtime.

That sounds good, but out of an abundance of caution, can we merge as-is, then make the feature default? I want to get a few days of no-issue CI builds and general usage. Also, I want to take another pass on the testing side, improving the one existing test and possibly including some other tests that others have been working on.

**Comment by alexcrichton** [comment]
Sure yeah, would you prefer to not test at all on CI while manual testing is done?

**Comment by abrown** [comment]
I'm fine with manual testing for now (like, a week?) of the wasi-threads parts; the parts I'm concerned about getting some usage of are the changes to wasi-common.

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 17]
```diff
@@ -188,11 +189,16 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
...
-            .await
+        if let Some(table) = self.table_mut() {
+            table
+                .get_file_mut(u32::from(fd))?
+                .get_cap_mut(FileCaps::FDSTAT_SET_FLAGS)?
+                .set_fdflags(FdFlags::from(flags))
+                .await
+        } else {
+            log::warn!("`fd_fdstat_set_flags` does not work with wasi-threads enabled; see https://github.com/bytecodealliance/wasmtime/issues/5643");
+            Err(Error::invalid_argument())
```
Comment (pchickey) [inline crates/wasi-common/src/snapshots/preview_1.rs:200]: bikeshed, but I think NOTSUP is a more appropriate error code for this
```suggestion
            Err(Error::not_supported())
```

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 18]
```diff
@@ -1112,7 +1115,7 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
             while copied < buf.len() {
                 let len = (buf.len() - copied).min(MAX_SHARED_BUFFER_SIZE as u32);
                 let mut tmp = vec![0; len as usize];
-                self.random.try_fill_bytes(&mut tmp)?;
+                self.random.lock().unwrap().try_fill_bytes(&mut tmp)?;
```
Comment (pchickey) [inline crates/wasi-common/src/snapshots/preview_1.rs:1118]: Could we return an error instead of panic if for some reason this lock is poisoned?

File: crates/wasi-common/src/snapshots/preview_1.rs [hunk 19]
```diff
@@ -1124,7 +1127,7 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
             // If the Wasm memory is non-shared, copy directly into the linear
             // memory.
             let mem = &mut buf.as_slice_mut()?.unwrap();
-            self.random.try_fill_bytes(mem)?;
+            self.random.lock().unwrap().try_fill_bytes(mem)?;
```
Comment (pchickey) [inline crates/wasi-common/src/snapshots/preview_1.rs:1130]: Same

File: crates/wasi/src/lib.rs [hunk 20]
```diff
@@ -82,3 +82,45 @@ pub mod snapshots {
 }
 }
 }
+
+/// Exit the process with a conventional OS error code as long as Wasmtime
+/// understands the error. If the error is not an `I32Exit` or `Trap`, return
+/// the error back to the caller for it to decide what to do.
+///
+/// Note: this function is designed for CLI usage of Wasmtime; embedders of
+/// Wasmtime may want different error handling.
```
Comment (pchickey) [inline crates/wasi/src/lib.rs:91]: I'd reword "CLI usage" to something like its "for usage where it is acceptable for wasmtime failures to terminate the parent process, such as in the wasmtime CLI", and say something like "not suitable for use in multi-tenant embeddings" as a more obvious warning? Could we hide the presence of this function behind an off-by-default cargo feature, to give more assurance to multi-tenant embeddings that they are safe from this?

File: crates/wasi-threads/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+                    "spawned thread id = {}; calling start function `{}` with: {}",
+                    wasi_thread_id,
+                    WASI_ENTRY_POINT,
+                    thread_start_arg
+                );
+                match thread_entry_point.call(&mut store, (wasi_thread_id, thread_start_arg)) {
+                    Ok(_) => log::trace!("exiting thread id = {} normally", wasi_thread_id),
+                    Err(e) => {
+                        log::trace!("exiting thread id = {} due to error", wasi_thread_id);
+                        let e = maybe_exit_on_error(e);
```
Comment (pchickey) [inline crates/wasi-threads/src/lib.rs:75]: Terminating the parent process is a really heavy hammer and makes this whole crate unsuitable for use in multi-tenant embeddings. It would be helpful to understand exactly why this crate had to make that design choice, and to please note it somewhere high level, like in a crate-level README

File: tests/all/cli_tests/threads.wat (hunk 4 above)
Comment (abrown) [reply tests/all/cli_tests/threads.wat]: Ended up doing this anyways in this PR to try to get CI to go green.

File: crates/wasi-common/src/snapshots/preview_1.rs (hunk 18 above)
Comment (abrown) [reply crates/wasi-common/src/snapshots/preview_1.rs:1118]: Sure, but which: `perm`? `invalid_argument`? `not_supported`? I don't know if any fit well.

Comment (abrown) [reply crates/wasi-common/src/snapshots/preview_1.rs:1118]: Also, if the `Mutex` ever does get poisoned (which doesn't seem likely to happen?) wouldn't we want to crash Wasmtime to figure out what is going on rather than letting the Wasm module proceed?

File: crates/wasi-threads/src/lib.rs (hunk 21 above)
Comment (abrown) [reply crates/wasi-threads/src/lib.rs:75]: Hopefully this hammer is only temporary; @alexcrichton and I had discussed some early thoughts on how we might shut down all of the threads when necessary. It is complex to do so, though, so providing experimental wasi-threads access sooner for some users seemed more valuable than solving the problem and waiting longer.

File: crates/wasi-common/src/snapshots/preview_1.rs (hunk 18 above)
Comment (pchickey) [reply crates/wasi-common/src/snapshots/preview_1.rs:1118]: Yeah, nevermind, I guess I didn't think this through - this is a per-ctx mutex so it would only get poisoned if another thread panicked already, so panicking more in the same store is fine.

**Review by pchickey — APPROVED** [review]

**Comment by abrown** [comment]
Rebased to handle conflicts due to version bumps and #5683.

**Comment by TerrorJack** [comment]
Note that the `thread-spawn` name in `wasi-libc` isn't coherent with the `thread_spawn` name used in `wasmtime`.

**Comment by abrown** [comment]
> Note that the `thread-spawn` name in `wasi-libc` isn't coherent with the `thread_spawn` name used in `wasmtime`.

@TerrorJack, thanks for the heads up. Yes, there are a few details that still need to be cleaned up, like this one. I opened #5748 to resolve this.

File: tests/all/cli_tests/threads.wat [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  (import "" "memory" (memory $shmem 1 1 shared))
+  (import "wasi_snapshot_preview1" "fd_write"
+    (func $__wasi_fd_write (param i32 i32 i32 i32) (result i32)))
+  (import "wasi_snapshot_preview1" "proc_exit"
+    (func $__wasi_proc_exit (param i32)))
+  (import "wasi" "thread_spawn"
+    (func $__wasi_thread_spawn (param i32) (result i32)))
+
+  (func (export "_start")
+    (local $i i32)
```
Comment (yamt) [inline tests/all/cli_tests/threads.wat:13]: unused local

File: tests/all/cli_tests/threads.wat [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+
+    ;; Wait for all the threads to notify us that they are done.
+    (loop $again
+      ;; Retrieve the i32 at address 128, compare it to -1 (it should always
+      ;; fail) and load it atomically to check if all three threads are
+      ;; complete. This wait is for 1ms or until notified, whichever is first.
+      (drop (memory.atomic.wait32 (i32.const 128) (i32.const -1) (i64.const 1000000)))
```
Comment (yamt) [inline tests/all/cli_tests/threads.wat:28]: if i read this correctly, this `wait` would always return 1 immediately without sleeping.

File: tests/all/cli_tests/threads.wat [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+    ;; After printing, we atomically increment the value at address 128 and then
+    ;; wake up the main thread's join loop.
+    (drop (i32.atomic.rmw.add (i32.const 128) (i32.const 1)))
+    (drop (memory.atomic.notify (i32.const 128) (i32.const 1)))
+  )
+
+  ;; A helper function for printing ptr-len strings.
+  (func $print (param $ptr i32) (param $len i32)
+    (i32.store (i32.const 8) (local.get $len))
+    (i32.store (i32.const 4) (local.get $ptr))
```
Comment (yamt) [inline tests/all/cli_tests/threads.wat:48]: this function doesn't seem thread-safe.

File: tests/all/cli_tests/threads.wat (hunk 23 above)
Comment (abrown) [reply tests/all/cli_tests/threads.wat:28]: Ah, yeah, this is a mistake due to a refactoring: originally I intended the `i` local (unused now) to keep track of the expected values but that was lost as I changed things. I'll add a note to fix this...

File: tests/all/cli_tests/threads.wat (hunk 24 above)
Comment (abrown) [reply tests/all/cli_tests/threads.wat:48]: Are you thinking about atomic stores or a lock around the `$__wasi_fd_write` call? (Recall that `wasi-common` has some internal locking).

Comment (yamt) [reply tests/all/cli_tests/threads.wat:48]: no. it's about the linear memory used for iovec.

File: tests/all/cli_tests/threads.wat (hunk 23 above)
Comment (abrown) [reply tests/all/cli_tests/threads.wat:28]: Fixed in #5858
